### PR TITLE
[stdlib] Add and use _identityCast, NFC

### PR DESCRIPTION
**Pull request by dabrahams**
@jckarter Did you add some `unsafeBitCast`s recently that would be better as `_identityCast`s?

### CODE REVIEW

**Comment by dabrahams** [comment]
@swift-ci Please smoke test and merge

File: stdlib/public/core/Builtin.swift [hunk 1]
```diff
@@ -114,6 +114,18 @@ public func unsafeBitCast<T, U>(_ x: T, to type: U.Type) -> U {
...
 
+/// Returns `x` as its concrete type `U`.
+///
+/// This cast can be useful for dispatching to specializations of generic
+/// functions.
+///
+/// - Requires: `x` has type `U`.
+@_transparent
+public func _identityCast<T, U>(_ x: T, to expectedType: U.Type) -> U {
+  _precondition(T.self == expectedType, "_identityCast to wrong type")
```
Comment (jrose-apple) [inline stdlib/public/core/Builtin.swift]: This is checking the static type rather than the dynamic type, no?

Comment (jrose-apple) [inline stdlib/public/core/Builtin.swift]: (as opposed to `type(of: x) == expectedType`)

Comment (dabrahams) [inline stdlib/public/core/Builtin.swift]: That's what I had before.  I'm not sure it matters for the use-cases in question though, and I thought the optimizer would have an easier time with this.

**Comment by dabrahams** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
There's this one: https://github.com/apple/swift/blob/master/stdlib/public/core/KeyPath.swift#L204 I thought I had more, but I guess I managed to refactor them away.

File: stdlib/public/core/Builtin.swift (hunk 1 above)
Comment (jrose-apple) [inline stdlib/public/core/Builtin.swift]: Oh, I guess it only matters for classes, where you wouldn't want to use this cast anyway.

**Pull request merged**